src/abstractdatatable/join.jl

```diff
@@ -162,31 +162,31 @@ DataTableJoiner{DT1<:AbstractDataTable, DT2<:AbstractDataTable}(dtl::DT1, dtr::D
     DataTableJoiner{DT1,DT2}(dtl, dtr, on)
 
 # helper map between the row indices in original and joined table
-immutable _RowIndexMap
+immutable RowIndexMap
     orig::Vector{Int} # row indices in the original table
     join::Vector{Int} # row indices in the resulting joined table
 end
 
-Base.length(x::_RowIndexMap) = length(x.orig)
+Base.length(x::RowIndexMap) = length(x.orig)
 
 # fix the result of the rightjoin by taking the nonnull values from the right table
 function fix_rightjoin_column!(res_col::AbstractArray, col_ix::Int, joiner::DataTableJoiner,
-                               all_orig_left_ixs::Vector{Int}, rightonly_ixs::_RowIndexMap)
+                               all_orig_left_ixs::Vector{Int}, rightonly_ixs::RowIndexMap)
     res_col[rightonly_ixs.join] = joiner.dtr_on[rightonly_ixs.orig, col_ix]
     res_col
 end
 
 # composes the joined data table using the maps between the left and right
 # table rows and the indices of rows in the result
 function compose_joined_table(joiner::DataTableJoiner,
-                left_ixs::_RowIndexMap, leftonly_ixs::_RowIndexMap,
-                right_ixs::_RowIndexMap, rightonly_ixs::_RowIndexMap)
+                              left_ixs::RowIndexMap, leftonly_ixs::RowIndexMap,
+                              right_ixs::RowIndexMap, rightonly_ixs::RowIndexMap)
     @assert length(left_ixs) == length(right_ixs)
     # compose left half of the result taking all left columns
     all_orig_left_ixs = vcat(left_ixs.orig, leftonly_ixs.orig)
     if length(leftonly_ixs) > 0
         # permute the indices to restore left table rows order
-        all_orig_left_ixs[[left_ixs.join; leftonly_ixs.join]] = all_orig_left_ixs
+        all_orig_left_ixs[vcat(left_ixs.join, leftonly_ixs.join)] = all_orig_left_ixs
     end
     left_dt = DataTable(Any[resize!(col[all_orig_left_ixs], length(all_orig_left_ixs)+length(rightonly_ixs))
```

Comment:
Here we are allocating `col[all_orig_left_ixs]` and then resizing it (by padding NULLs) once more. I wonder if it's possible for Categorical/NullableArrays to do it in one operation. However, `col` might be non-nullable (I don't know if it is allowed in DataTables but I'm very much for allowing non-nullable columns because it allows to enforce data integrity when needed), then the logic is more complicated:
  1) for right-joins the values of the "on" column (i.e. used for joining) are to be taken from the right table (that's what is done later by `fix_rightjoin_column!()`, but for non-nullable columns it has to be done here
  2) for full-joins adding nulls to non-nullable "on" columns should generate an error
  3) an error should be raised if it's not the "on" column and is not nullable